collect_app/src/main/java/org/odk/collect/android/external/FormUriActivity.kt

```diff
@@ -87,15 +100,46 @@ class FormUriActivity : ComponentActivity() {
         DaggerUtils.getComponent(this).inject(this)
         setContentView(R.layout.circular_progress_indicator)
 
-        formUriViewModel.error.observe(this) {
-            if (it != null) {
-                displayErrorDialog(it)
-            } else if (savedInstanceState?.getBoolean(FORM_FILLING_ALREADY_STARTED) != true) {
-                startForm()
+        if (savedInstanceState?.getBoolean(FORM_FILLING_ALREADY_STARTED) == true) {
+            return
+        }
+
+        formUriViewModel.formInspectionResult.observe(this) {
+            when (it) {
+                is FormInspectionResult.Error -> displayErrorDialog(it.error)
+                is FormInspectionResult.Savepoint -> displaySavePointRecoveryDialog(it.savepoint)
+                is FormInspectionResult.Valid -> startForm()
             }
         }
     }
 
+    private fun displaySavePointRecoveryDialog(savepoint: Savepoint) {
+        MaterialAlertDialogBuilder(this)
+            .setTitle(string.savepoint_recovery_dialog_title)
+            .setMessage(SimpleDateFormat(getString(string.savepoint_recovery_dialog_message), Locale.getDefault()).format(File(savepoint.savepointFilePath).lastModified()))
+            .setPositiveButton(string.recover) { _, _ ->
+                val uri = intent.data!!
+                val uriMimeType = contentResolver.getType(uri)!!
+                if (uriMimeType == FormsContract.CONTENT_ITEM_TYPE) {
+                    intent.data = FormsContract.getUri(projectsDataService.getCurrentProject().uuid, savepoint.formDbId)
+                }
+                startForm()
+            }
+            .setNegativeButton(string.do_not_recover) { _, _ ->
+                scheduler.immediate(
+                    background = {
+                        savepointsRepositoryProvider.get().delete(savepoint.formDbId, savepoint.instanceDbId)
```

Comment:
> I'm guessing this line is required so that the save point doesn't get loaded by FormLoaderTask later on in FormFillingActivity

Yes, and it also should be removed so that you don't see the recovery dialog when you start the same again.
I've added tests in `FormUriActivityTest`.